File: Algorithms/0938_Range_Sum_of_BST/Java/RangeSumofBSTSolution2.java

// Space: O(1)
// Time: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int rangeSumBST(TreeNode root, int low, int high) {
        if (root == null) return 0;
        if (root.left == null && root.right == null && low <= root.val && root.val <= high) return root.val;
        return dfs(root,low,high);
    }

    private int dfs(TreeNode root, int low, int high) {
        if (root == null) return 0;
        if (root.val<low) return dfs(root.right,low,high);
        if (root.val>high) return dfs(root.left,low,high);
        return dfs(root.left,low,high) + root.val + dfs(root.right,low,high);
    }
}

File: Algorithms/0938_Range_Sum_of_BST/Python/Range_Sum_of_BST_Solution_2.py
# Space: O(1)
# Time: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root, low, high):
        if root.left is None and root.right is None: return root.val

        def dfs(root, low, high):
            if root is None: return 0

            if root.val < low:
                return dfs(root.right, low, high)
            if root.val > high:
                return dfs(root.left, low, high)

            return dfs(root.left, low, high) + root.val + dfs(root.right, low, high)

        return dfs(root, low, high)





                                                                                                